exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java

```diff
@@ -387,62 +443,100 @@ private InfoSnapshot makeTargetInfo(Resource resource) {
    *     {@code true}.
    * @param attributes the attributes to be converted.
    * @param additionalAttributes optional list of key/value pairs, may be empty.
    */
+  @SuppressWarnings({"rawtypes", "unchecked"})
   private Labels convertAttributes(
+      @Nullable Attributes resourceAttributes,
       @Nullable InstrumentationScopeInfo scope,
       Attributes attributes,
       String... additionalAttributes) {
-    int numberOfScopeAttributes = 0;
-    if (otelScopeEnabled && scope != null) {
-      numberOfScopeAttributes = scope.getVersion() == null ? 1 : 2;
-    }
-    String[] names =
-        new String[attributes.size() + numberOfScopeAttributes + additionalAttributes.length / 2];
-    String[] values = new String[names.length];
-    int[] pos = new int[] {0}; // using an array because we want to increment in a forEach() lambda.
+
+    List<AttributeKey> allowedAttributeKeys =
+        addResourceAttributesAsLabels
+            ? filterAllowedResourceAttributeKeys(resourceAttributes)
+            : Collections.emptyList();
+
+    ArrayList<String> names = new ArrayList<>();
+    ArrayList<String> values = new ArrayList<>();
     attributes.forEach(
         (key, value) -> {
-          names[pos[0]] = sanitizeLabelName(key.getKey());
-          values[pos[0]] = value.toString();
-          pos[0]++;
+          names.add(sanitizeLabelName(key.getKey()));
+          values.add(value.toString());
         });
+
     for (int i = 0; i < additionalAttributes.length; i += 2) {
-      names[pos[0]] = additionalAttributes[i];
-      values[pos[0]] = additionalAttributes[i + 1];
-      pos[0]++;
+      names.add(requireNonNull(additionalAttributes[i]));
+      values.add(requireNonNull(additionalAttributes[i + 1]));
     }
+
     if (otelScopeEnabled && scope != null) {
-      names[pos[0]] = OTEL_SCOPE_NAME;
-      values[pos[0]] = scope.getName();
-      pos[0]++;
+      names.add(OTEL_SCOPE_NAME);
+      values.add(scope.getName());
       if (scope.getVersion() != null) {
-        names[pos[0]] = OTEL_SCOPE_VERSION;
-        values[pos[0]] = scope.getVersion();
-        pos[0]++;
+        names.add(OTEL_SCOPE_VERSION);
+        values.add(scope.getVersion());
       }
     }
+
+    if (resourceAttributes != null) {
+      for (AttributeKey attributeKey : allowedAttributeKeys) {
+        // Don't write the same attribute twice.
+        if (names.stream().anyMatch(name -> name.equals(attributeKey.getKey()))) {
+          continue;
+        }
+        Object attributeValue = resourceAttributes.get(attributeKey);
+        if (attributeValue != null) {
+          names.add(sanitizeLabelName(attributeKey.getKey()));
+          values.add(attributeValue.toString());
+        }
+      }
+    }
+
     return Labels.of(names, values);
   }
 
+  @SuppressWarnings("rawtypes")
+  private List<AttributeKey> filterAllowedResourceAttributeKeys(@Nullable Attributes attributes) {
```

Comment:
```suggestion
  private List<AttributeKey<?>> filterAllowedResourceAttributeKeys(@Nullable Attributes attributes) {
```

There's no reason we can have a wildcard type here to get rid of the `@SuppressWarnings("rawtypes")`